### PR TITLE
Allocator tests

### PR DESCRIPTION
**Pull request by Przemog1**

### CODE REVIEW

**Comment by devshgraphicsprogramming** [comment]
merge other examples into ex 10

File: examples_tests/43.AllocatorTest/main.cpp [hunk 1]
```diff
@@ -1,53 +1,219 @@
...
+using namespace core;
 
-/*
-	Problems with GeneralPurposeallocator:
-		1. GeneralPurposeallocator is not able to allocate as much address space as it is expected to
-		2. Program doesn't compile, when I try to call safe_shrink_size
-		3. "Assertion failed: false, file *\include\irr\core\alloc\GeneralpurposeAddressAllocator.h, line 368"
+/*
+	If traits say address allocator supports arbitrary order free,
+	then ETAT_CHOOSE_RANDOMLY is valid. Otherwise ETAT_CHOOSE_MOST_RECENTLY_ALLOCATED
```
Comment (devshgraphicsprogramming) [inline examples_tests/43.AllocatorTest/main.cpp:11]: stack allocator is "most recently allocated" but the linear allocator has `free_addr` as a NO-OP, basically the whole allocator needs to be `reset` to actually reclaim the memory.

File: examples_tests/43.AllocatorTest/main.cpp [hunk 2]
```diff
@@ -1,53 +1,219 @@
...
+constexpr size_t maxAlignOffset = 64;
+constexpr size_t maxBlockSize = 600;
+
+//TODO: full static
+class RandomNumberGenerator
+{
+public:
+	inline uint32_t getRndAllocCnt()    { return  allocsPerFrameRange(mt);  }
+	inline uint32_t getRndSize()        { return  sizePerFrameRange(mt); }
+	inline uint32_t getRndMaxAlign()    { return  (1u << maxAlignmentExpPerFrameRange(mt)); } //128 is max
```
Comment (devshgraphicsprogramming) [inline examples_tests/43.AllocatorTest/main.cpp:43]: max alignment and size do not really apply to the Pool Address allocator, because there the allocations must be a fixed size (block size) so the allocation and size will be both equal to that.

You can however, test different block sizes for multiple instances of Pool Address Allocators.

File: examples_tests/43.AllocatorTest/main.cpp [hunk 3]
```diff
@@ -1,53 +1,219 @@
...
+struct AllocationCreationParameters
+{	
+	size_t multiAllocCnt;              //! Specifies amount of adress to be allocated with certain choosen allocator
+	size_t adressesToDeallocateCnt;    //! Specifies amount of adress to be deallocated with certain choosen allocator. Must be less than all allocated and must pick number less than `traits::max_multi_free`, but we don't have `max_multi_free`
+};
+
+constexpr size_t minAllocs = 10000u;
+constexpr size_t maxAllocs = 20000u;
+constexpr size_t maxSizeForSquaring = 1024u;
+constexpr size_t maxAlignmentExp = 6u;
```
Comment (devshgraphicsprogramming) [inline examples_tests/43.AllocatorTest/main.cpp:30]: oooh, alignment can be muuuch bigger than that (at least 4096)

File: examples_tests/43.AllocatorTest/main.cpp (hunk 2 above)
Comment (devshgraphicsprogramming) [reply examples_tests/43.AllocatorTest/main.cpp:43]: Also I don't think there's really a requirement for my Address Allocators to only service PoT alignments, in-fact the alignments can be any non-zero positive integer.... this is why GeneralPurposeAddressAllocator can be used to manage any format mesh vertices ;)

File: examples_tests/43.AllocatorTest/main.cpp [hunk 4]
```diff
@@ -1,53 +1,219 @@
...
+constexpr size_t minByteSizeForReservingMemory = 2048;			// 2kB
+constexpr size_t maxByteSizeForReservingMemory = 2147483648;	// 2GB
```
Comment (devshgraphicsprogramming) [inline examples_tests/43.AllocatorTest/main.cpp:32]: the address allocator's `reserved_size` tells you exactly how much reserved memory you need.

File: examples_tests/43.AllocatorTest/main.cpp [hunk 5]
```diff
@@ -1,53 +1,219 @@
...
+{
+	ETAT_CHOOSE_RANDOMLY,
+	ETAT_CHOOSE_MOST_RECENTLY_ALLOCATED,
+	ETAT_COUNT
+};
+
+struct AllocationCreationParameters
+{	
+	size_t multiAllocCnt;              //! Specifies amount of adress to be allocated with certain choosen allocator
+	size_t adressesToDeallocateCnt;    //! Specifies amount of adress to be deallocated with certain choosen allocator. Must be less than all allocated and must pick number less than `traits::max_multi_free`, but we don't have `max_multi_free`
```
Comment (devshgraphicsprogramming) [inline examples_tests/43.AllocatorTest/main.cpp:24]: the number of multi_allocs and how many addresses each multi_alloc should allocate should be random.

File: examples_tests/43.AllocatorTest/main.cpp [hunk 6]
```diff
@@ -1,53 +1,219 @@
...
-				os::Printer::print(std::to_string(addr));
+	}
+	
+	//struct for holding temporary data (inputs of `multi_alloc_addr` i `multi_free_addr`)
+	struct EntryForFrameData
+	{
+		uint32_t outAddr[maxAllocs] = { ~0u };
+		uint32_t sizes[maxAllocs] = { 0u };
+		uint32_t alignments[maxAllocs] = { 0u };
+	} perFrameData;
```
Comment (devshgraphicsprogramming) [inline examples_tests/43.AllocatorTest/main.cpp:84]: correctness, not performance is the focus here, so you have my blessing to use `std::vector`

File: examples_tests/43.AllocatorTest/main.cpp [hunk 7]
```diff
@@ -1,53 +1,219 @@
...
-			if (addr == alctr.invalid_address)
-				os::Printer::print("invalid adress");
-			else
-				os::Printer::print(std::to_string(addr));
+	}
+	
+	//struct for holding temporary data (inputs of `multi_alloc_addr` i `multi_free_addr`)
+	struct EntryForFrameData
+	{
+		uint32_t outAddr[maxAllocs] = { ~0u };
```
Comment (devshgraphicsprogramming) [inline examples_tests/43.AllocatorTest/main.cpp:81]: initialize all with `invalid_address`, the value ~0u is not that, `multi_alloc` wont work is the output array of addresses is not valid.

File: examples_tests/43.AllocatorTest/main.cpp [hunk 8]
```diff
@@ -1,53 +1,219 @@
...
+		uint32_t sizes[maxAllocs] = { 0u };
+		uint32_t alignments[maxAllocs] = { 0u };
+	} perFrameData;
+	
+	void executeAllocatorTest()
+	{
+		os::Printer::log("Executing Pool Allocator Test!", ELL_INFORMATION);
+	
+		for (size_t i = 0; i < creationParameters.multiAllocCnt; ++i)
+			executeForFrame();
```
Comment (devshgraphicsprogramming) [inline examples_tests/43.AllocatorTest/main.cpp:91]: you need 3 scopes:
- random number of iteration loops (times `reset` of an allocator is performed)
- random number of multi-alloc calls
- random number of allocations in the multi-alloc

You need to handle failure to allocate gracefully.

File: examples_tests/43.AllocatorTest/main.cpp [hunk 9]
```diff
@@ -1,53 +1,219 @@
...
+
+		// randomly decide the number of iterations of allocation and reset
+			// declare `core::vector` to hold allocated addresses and sizes
+
+			// randomly decide how many `multi_allocs` to do
+				// randomly decide how many allocs in a `multi_alloc` NOTE: must pick number less than `traits::max_multi_alloc`
+					// randomly decide sizes (but always less than `address_allocator_traits::max_size`)
+				// record all successful alloc addresses to the `core::vector`
+
+				// run random dealloc function
```
Comment (devshgraphicsprogramming) [inline examples_tests/43.AllocatorTest/main.cpp:110]: its important to throw in some random multi frees between every multi-alloc, this helps us test fragmentation and bookeeping of free memory.

Simply doing all allocations followed by all frees would leave bug loopholes.

File: examples_tests/43.AllocatorTest/main.cpp [hunk 10]
```diff
@@ -1,53 +1,219 @@
...
+				// reset
+			// ELSE
+				// free everything with a series of multi_free
+	void executeForFrame()
+	{
+		uint32_t randMaxAlign = rng.getRndMaxAlign();
+		uint32_t randAddressSpaceSize = rng.getRndBuffSize();
+		uint32_t randAlignOffset = rng.getRandomNumber(0u, randMaxAlign - 1u);
+		uint32_t randOffset = rng.getRandomNumber(0u, randAddressSpaceSize - 1u); //?
+		uint32_t randBlockSz = rng.getRandomNumber(0u, randAddressSpaceSize - 1u); //?
```
Comment (devshgraphicsprogramming) [inline examples_tests/43.AllocatorTest/main.cpp:125]: looks good, except that I'd make the block size between 1u and `(randAddressSpaceSize-randOffset)/2u`

Comment (devshgraphicsprogramming) [reply examples_tests/43.AllocatorTest/main.cpp:125]: and assertthe end result is >0

File: examples_tests/43.AllocatorTest/main.cpp [hunk 11]
```diff
@@ -1,53 +1,219 @@
...
+
+		uint32_t allocationAmountForSingeFrameTest = rng.getRndAllocCnt();
+		
+		const auto reservedSize = PoolAddressAllocator::reserved_size(randMaxAlign, randAddressSpaceSize, randBlockSz); // 3rd parameter onward is custom for each address alloc type
+		void* reservedSpace = _IRR_ALIGNED_MALLOC(reservedSize, _IRR_SIMD_ALIGNMENT);
+		auto poolAlctr = PoolAddressAllocator(reservedSpace, randOffset, randAlignOffset, randMaxAlign, randAddressSpaceSize, randBlockSz);
+
+		// randomly decide how many `multi_allocs` to do
+		const uint32_t multiAllocCnt = rng.getRandomNumber(1u, 5u); //range?
```
Comment (devshgraphicsprogramming) [inline examples_tests/43.AllocatorTest/main.cpp:134]: as many as reasonable time-wise for the test, probably in the thousands or millions though.

File: examples_tests/43.AllocatorTest/main.cpp [hunk 12]
```diff
@@ -1,53 +1,219 @@
...
+		for (uint32_t i = 0u; i < multiAllocCnt; i++)
+		{
+			// randomly decide how many allocs in a `multi_alloc` NOTE: must pick number less than `traits::max_multi_alloc`
+			const uint32_t allocCntInMultiAlloc = rng.getRandomNumber(1u, 4096u); //?
```
Comment (devshgraphicsprogramming) [inline examples_tests/43.AllocatorTest/main.cpp:138]: yeah, all good, just use the max multi alloc trait

File: examples_tests/43.AllocatorTest/main.cpp [hunk 13]
```diff
@@ -1,53 +1,219 @@
...
+
+			for (size_t i = 0u; i < allocCntInMultiAlloc; ++i)
+			{
+				// randomly decide sizes (but always less than `address_allocator_traits::max_size`)
+				perFrameData.sizes[i] = rng.getRandomNumber(1u, Traits::max_size(poolAlctr));
+				perFrameData.alignments[i] = rng.getRndMaxAlign();
```
Comment (devshgraphicsprogramming) [inline examples_tests/43.AllocatorTest/main.cpp:144]: pool allocator cant do random sizes and alignments, they always need to be block sized... other allocators can though.

File: examples_tests/43.AllocatorTest/main.cpp [hunk 14]
```diff
@@ -1,53 +1,219 @@
...
+		_IRR_ALIGNED_FREE(reservedSpace);
+	}
+
+	// random dealloc function
+		// randomly decide how many calls to `multi_alloc` | ??
+			// randomly how many addresses we should deallocate (but obvs less than all allocated) NOTE: must pick number less than `traits::max_multi_free`
+				// if traits say address allocator supports arbitrary order free, then choose randomly, else choose most recently allocated
+	void randFreeAllocatedAddresses(PoolAddressAllocator& alctr)
+	{
+		//here I'm freeing previously generated addresses, is it ok?
```
Comment (devshgraphicsprogramming) [inline examples_tests/43.AllocatorTest/main.cpp:178]: yes ofc.

File: examples_tests/43.AllocatorTest/main.cpp [hunk 15]
```diff
@@ -1,53 +1,219 @@
...
+			// free everything with a series of multi_free
+			while (results.size() != 0u)
+				randFreeAllocatedAddresses(poolAlctr);
+		}
+		 
+		_IRR_ALIGNED_FREE(reservedSpace);
+	}
+
+	// random dealloc function
+		// randomly decide how many calls to `multi_alloc` | ??
```
Comment (devshgraphicsprogramming) [inline examples_tests/43.AllocatorTest/main.cpp:173]: typo.. I meant multi-free

File: examples_tests/43.AllocatorTest/main.cpp [hunk 16]
```diff
@@ -1,53 +1,219 @@
...
+		const uint32_t addressesToFreeCnt = rng.getRandomNumber(0u, results.size());
+
+		for (uint32_t i = 0u; i < addressesToFreeCnt; i++)
+		{
+			perFrameData.outAddr[i] = results[i].outAddr;
+			perFrameData.sizes[i] = results[i].size;
 		}
-		_IRR_ALIGNED_FREE(resSpc);
+
+		Traits::multi_free_addr(alctr, addressesToFreeCnt, perFrameData.outAddr, perFrameData.sizes);
```
Comment (devshgraphicsprogramming) [inline examples_tests/43.AllocatorTest/main.cpp:193]: you want a random number of calls to this.

File: examples_tests/43.AllocatorTest/main.cpp [hunk 17]
```diff
@@ -1,53 +1,219 @@
...
+				// if traits say address allocator supports arbitrary order free, then choose randomly, else choose most recently allocated
+	void randFreeAllocatedAddresses(PoolAddressAllocator& alctr)
+	{
+		//here I'm freeing previously generated addresses, is it ok?
+
+		if(results.size() == 0u)
+			return;
+
+		// randomly how many addresses we should deallocate (but obvs less than all allocated) NOTE: must pick number less than `traits::max_multi_free`
+		//TODO: random interval
```
Comment (devshgraphicsprogramming) [inline examples_tests/43.AllocatorTest/main.cpp:184]: you mean the traits max multi free

File: examples_tests/43.AllocatorTest/main.cpp [hunk 18]
```diff
@@ -1,53 +1,219 @@
...
+		const uint32_t addressesToFreeCnt = rng.getRandomNumber(0u, results.size());
+
+		for (uint32_t i = 0u; i < addressesToFreeCnt; i++)
+		{
+			perFrameData.outAddr[i] = results[i].outAddr;
+			perFrameData.sizes[i] = results[i].size;
```
Comment (devshgraphicsprogramming) [inline examples_tests/43.AllocatorTest/main.cpp:190]: erm actually....

You either want to:
- For a Stack allocator, add the results in reverse (FILO) and remove them from the vector.
- For all other allocator types randomly shuffle the `results`, and then do as above (right now you're always freeing in the same order they were allocated, which is not a good test). Or do some other method to **randomly choose** N items from a set of K.

File: examples_tests/43.AllocatorTest/main.cpp [hunk 19]
```diff
@@ -1,53 +1,219 @@
...
 		}
-		_IRR_ALIGNED_FREE(resSpc);
+
+		Traits::multi_free_addr(alctr, addressesToFreeCnt, perFrameData.outAddr, perFrameData.sizes);
+		results.erase(results.begin(), results.begin() + addressesToFreeCnt);
```
Comment (devshgraphicsprogramming) [inline examples_tests/43.AllocatorTest/main.cpp:194]: The O(N^2) nature of this `erase` call will cripple us.

Comment (devshgraphicsprogramming) [reply examples_tests/43.AllocatorTest/main.cpp:194]: You need:
- a dummy (NOOP) container for linear allocator
- a FILO container for stack allocator
- a container that supports random deletion and in-order insert in O(logn) or O(1)

File: examples_tests/43.AllocatorTest/main.cpp [hunk 20]
```diff
@@ -1,53 +1,219 @@
...
+
+			// record all successful alloc addresses to the `core::vector`
+			if (perFrameData.outAddr[i] != PoolAddressAllocator::invalid_address)
+				results.push_back({ perFrameData.outAddr[i], perFrameData.sizes[i], perFrameData.alignments[i] });
+
+			randFreeAllocatedAddresses(poolAlctr);
+		}
+
+		// randomly choose between reset and freeing all `core::vector` elements
+		bool reset = static_cast<bool>(rng.getRandomNumber(0u, 1u));
```
Comment (devshgraphicsprogramming) [inline examples_tests/43.AllocatorTest/main.cpp:157]: for a linear allocator, it would always need to be reset (since free is a noop)

File: examples_tests/43.AllocatorTest/main.cpp (hunk 2 above)
Comment (Przemog1) [reply examples_tests/43.AllocatorTest/main.cpp:43]: > Also I don't think there's really a requirement for my Address Allocators to only service PoT alignments
https://github.com/Devsh-Graphics-Programming/Nabla/blob/master/include/irr/core/alloc/AddressAllocatorBase.h#L73

Comment (devshgraphicsprogramming) [reply examples_tests/43.AllocatorTest/main.cpp:43]: that's max requestable that needs to be PoT, not the individuals.

File: examples_tests/43.AllocatorTest/main.cpp (hunk 13 above)
Comment (devshgraphicsprogramming) [reply examples_tests/43.AllocatorTest/main.cpp:144]: also you need to choose the rand alignment differently than the max for the allocator
https://github.com/Devsh-Graphics-Programming/Nabla/pull/1/files/0078e583bc8f69dd8697b9a6d5eb5e35f0aaa769#r496261998

File: examples_tests/43.AllocatorTest/main.cpp (hunk 4 above)
Comment (Przemog1) [reply examples_tests/43.AllocatorTest/main.cpp:32]: buffer size of a virtual memory comes from this range: 
<`minByteSizeForReservingMemory`, `maxByteSizeForReservingMemory`]>

names of those variables may be misleading

File: examples_tests/43.AllocatorTest/main.cpp (hunk 12 above)
Comment (Przemog1) [reply examples_tests/43.AllocatorTest/main.cpp:138]: I can't find this function :/

File: examples_tests/43.AllocatorTest/main.cpp (hunk 4 above)
Comment (devshgraphicsprogramming) [reply examples_tests/43.AllocatorTest/main.cpp:32]: ah ok, change the names then.

File: examples_tests/43.AllocatorTest/main.cpp (hunk 12 above)
Comment (devshgraphicsprogramming) [reply examples_tests/43.AllocatorTest/main.cpp:138]: https://github.com/Devsh-Graphics-Programming/Nabla/blob/master/include/irr/core/alloc/address_allocator_traits.h#L163

its a constexpr.

File: examples_tests/43.AllocatorTest/main.cpp (hunk 19 above)
Comment (Przemog1) [reply examples_tests/43.AllocatorTest/main.cpp]: > The O(N^2) nature of this erase call will cripple us.

isn't it linear? https://en.cppreference.com/w/cpp/container/vector/erase

Comment (devshgraphicsprogramming) [reply examples_tests/43.AllocatorTest/main.cpp]: for a single removal its linear O(N), now do N removals on a vector of N items will give you N^2 runtime

File: examples_tests/43.AllocatorTest/main.cpp [hunk 21]
```diff
@@ -269,4 +276,202 @@ int main()
...
+				os::Printer::print(std::to_string(addr));
+		}
+		_IRR_ALIGNED_FREE(resSpc);
+	}
+
+	//2
+	{
+		void* resSpc = _IRR_ALIGNED_MALLOC(core::GeneralpurposeAddressAllocator<uint32_t>::reserved_size(alignof(uint32_t), 16u, 1u), _IRR_SIMD_ALIGNMENT);
+		auto alctr = core::GeneralpurposeAddressAllocator<uint32_t>(resSpc, 0u, 0u, alignof(uint32_t), 16u, 1u);
+		alctr.safe_shrink_size(8u);
```
Comment (devshgraphicsprogramming) [inline examples_tests/43.AllocatorTest/main.cpp]: this code you can nuke, yours replaces it entirely.

File: examples_tests/10.AllocatorTest/main.cpp [hunk 22]
```diff
@@ -0,0 +1,478 @@
...
+#define ALLOCATOR_TEST
+//#define ALLOCATOR_TEST_OLD
+//#define ADDRESS_ALLOCATOR_TRAITS_TEST
+//#define ALLOC_PREF_TEST
+
+#ifdef ALLOCATOR_TEST
+
+struct AllocationCreationParameters
+{	
+	size_t multiAllocCnt;              //! Specifies amount of adress to be allocated with certain choosen allocator
```
Comment (devshgraphicsprogramming) [inline examples_tests/10.AllocatorTest/main.cpp:18]: call this something else its confusing, you have two multi alloc counts.

File: examples_tests/10.AllocatorTest/main.cpp [hunk 23]
```diff
@@ -0,0 +1,478 @@
...
+
+		if constexpr (std::is_same<AlctrType, core::LinearAddressAllocator<uint32_t>>::value)
+		{
+			alctr = AlctrType(nullptr, randAllocParams.offset, randAllocParams.alignOffset, randAllocParams.maxAlign, randAllocParams.addressSpaceSize);
+		}
+		else
+		{
+			const auto reservedSize = AlctrType::reserved_size(randAllocParams.maxAlign, randAllocParams.addressSpaceSize, randAllocParams.blockSz);
+			reservedSpace = _IRR_ALIGNED_MALLOC(reservedSize, _IRR_SIMD_ALIGNMENT);
+			alctr = AlctrType(reservedSpace, randAllocParams.offset, randAllocParams.alignOffset, randAllocParams.maxAlign, randAllocParams.addressSpaceSize, randAllocParams.blockSz);
```
Comment (devshgraphicsprogramming) [inline examples_tests/10.AllocatorTest/main.cpp:110]: for pool allocator the block size is basically a constant allocation size (can never be different)

But for the Generalpurpose allocator, its the minimum allocation size.

File: examples_tests/10.AllocatorTest/main.cpp [hunk 24]
```diff
@@ -0,0 +1,478 @@
...
+		}
+
+		// randomly decide how many `multi_allocs` to do
+		const uint32_t multiAllocCnt = rng.getRandomNumber(1u, 5u); //TODO: will change it later
+		for (uint32_t i = 0u; i < multiAllocCnt; i++)
+		{
+			outAddresses.clear();
+			sizes.clear();
+			alignments.clear();
```
Comment (devshgraphicsprogramming) [inline examples_tests/10.AllocatorTest/main.cpp:119]: this is wrong, it means you deallocate all allocations from each multi_alloc (which means you'll never multi_alloc on a fragmented allocator)

File: examples_tests/10.AllocatorTest/main.cpp [hunk 25]
```diff
@@ -0,0 +1,478 @@
...
+
+			// randomly decide how many allocs in a `multi_alloc` NOTE: must pick number less than `traits::max_multi_alloc`
+			constexpr uint32_t upperBound = core::address_allocator_traits<AlctrType>::maxMultiOps;
+			const uint32_t allocCntInMultiAlloc = rng.getRandomNumber(1u, upperBound);
+
+			for (size_t i = 0u; i < allocCntInMultiAlloc; ++i)
+			{
+				// randomly decide sizes (but always less than `address_allocator_traits::max_size`)
+				outAddresses.emplace_back(AlctrType::invalid_address);
```
Comment (devshgraphicsprogramming) [inline examples_tests/10.AllocatorTest/main.cpp:128]: resize + std::fill would be a good choice here

File: examples_tests/10.AllocatorTest/main.cpp [hunk 26]
```diff
@@ -0,0 +1,478 @@
...
+
+				if constexpr (std::is_same<AlctrType, core::PoolAddressAllocator<uint32_t>>::value)
+				{
+					sizes.emplace_back(randAllocParams.blockSz);
+					alignments.emplace_back(randAllocParams.blockSz);
+				}
+				else
+				{
+					sizes.emplace_back(rng.getRandomNumber(1u, Traits::max_size(alctr)));
+					alignments.emplace_back(rng.getRndMaxAlign());
```
Comment (devshgraphicsprogramming) [inline examples_tests/10.AllocatorTest/main.cpp:138]: for the generalpurpose allocator the size needs to be >= minimum size (block size)

For Pool allocator both size and alignment should be equal.

I dont recall any restrictions on alignment for generalpurpose allocator though.

Comment (devshgraphicsprogramming) [reply examples_tests/10.AllocatorTest/main.cpp:138]: its probably why generalpurpose crashes during the test.

File: examples_tests/10.AllocatorTest/main.cpp (hunk 24 above)
Comment (devshgraphicsprogramming) [reply examples_tests/10.AllocatorTest/main.cpp:119]: my bad, it got this confused with `results`

File: examples_tests/10.AllocatorTest/main.cpp (hunk 25 above)
Comment (devshgraphicsprogramming) [reply examples_tests/10.AllocatorTest/main.cpp:128]: actually you could resize all arrays to `core::address_allocator_traits<AlctrType>::maxMultiOps` in the constructor up-front (or even use a std:array)

File: examples_tests/10.AllocatorTest/main.cpp [hunk 27]
```diff
@@ -0,0 +1,478 @@
...
+	{
+		if(results.size() == 0u)
+			return;
+
+		outAddresses.clear();
+		sizes.clear();
+		alignments.clear();
+
+		// randomly decide how many calls to `multi_free`
+		const uint32_t multiFreeCnt = rng.getRandomNumber(1u, 10000u); //TODO
```
Comment (devshgraphicsprogramming) [inline examples_tests/10.AllocatorTest/main.cpp:189]: 1 to `results.size()` ?

File: examples_tests/10.AllocatorTest/main.cpp [hunk 28]
```diff
@@ -0,0 +1,478 @@
...
+
+		//TODO:
+		//shuffle results
+
+		for (uint32_t i = 0u; i < multiFreeCnt; i++)
+		{
+			// randomly how many addresses we should deallocate (but obvs less than all allocated) NOTE: must pick number less than `traits::max_multi_free`
+			const uint32_t addressesToFreeCnt = rng.getRandomNumber(0u, results.size()); //that should be restrained somehow, I think... so it it less likely to free all of allocated addresses
```
Comment (devshgraphicsprogramming) [inline examples_tests/10.AllocatorTest/main.cpp:197]: upper bound should be `min(core::address_allocator_traits<AlctrType>::maxMultiOps,1)`

Comment (devshgraphicsprogramming) [reply examples_tests/10.AllocatorTest/main.cpp:197]: dont restrain any more, we need to be able to hit mutli_free everything every once in a while, this could be an important edge case to test.

File: examples_tests/10.AllocatorTest/main.cpp [hunk 29]
```diff
@@ -0,0 +1,478 @@
...
+
+		outAddresses.clear();
+		sizes.clear();
+		alignments.clear();
+
+		// randomly decide how many calls to `multi_free`
+		const uint32_t multiFreeCnt = rng.getRandomNumber(1u, 10000u); //TODO
+
+		//TODO:
+		//shuffle results
```
Comment (devshgraphicsprogramming) [inline examples_tests/10.AllocatorTest/main.cpp:192]: only for generalpurpose.

File: examples_tests/10.AllocatorTest/main.cpp [hunk 30]
```diff
@@ -0,0 +1,478 @@
...
+			const uint32_t addressesToFreeCnt = rng.getRandomNumber(0u, results.size()); //that should be restrained somehow, I think... so it it less likely to free all of allocated addresses
+
+			for (uint32_t i = 0u; i < addressesToFreeCnt; i++)
+			{
+				outAddresses.push_back(results[i].outAddr);
+				sizes.push_back(results[i].size);
+			}
+
+			Traits::multi_free_addr(alctr, addressesToFreeCnt, outAddresses.data(), sizes.data());
+			results.erase(results.begin(), results.begin() + addressesToFreeCnt);
```
Comment (devshgraphicsprogramming) [inline examples_tests/10.AllocatorTest/main.cpp:206]: I'd rather you free the `results.end()-addressesToFreeCnt,results.end()` range because then you wont get the O(n^2) removal.

Comment (devshgraphicsprogramming) [reply examples_tests/10.AllocatorTest/main.cpp:206]: https://github.com/Devsh-Graphics-Programming/Nabla/pull/1#discussion_r496272644

File: examples_tests/10.AllocatorTest/main.cpp [hunk 31]
```diff
@@ -0,0 +1,438 @@
...
+				alctr.reset();
+				results.clear();
+			}
+			else
+			{
+				// free everything with a series of multi_free
+				while (results.size() != 0u)
+					randFreeAllocatedAddresses(alctr);
+			}
+		}
```
Comment (devshgraphicsprogramming) [inline examples_tests/10.AllocatorTest/main.cpp:138]: https://github.com/Devsh-Graphics-Programming/Nabla/pull/1#discussion_r496264837

> random number of iteration loops (times reset of an allocator is performed)
random number of multi-alloc calls with random number of allocations in the multi-alloc
random number of multi-free calls (not every multi-allocation is paired 1:1 with multi-free)

File: examples_tests/10.AllocatorTest/main.cpp (hunk 23 above)
Comment (devshgraphicsprogramming) [reply examples_tests/10.AllocatorTest/main.cpp]: STILL NOT DONE

File: examples_tests/10.AllocatorTest/main.cpp [hunk 32]
```diff
@@ -0,0 +1,420 @@
...
+		if constexpr (std::is_same<AlctrType, core::LinearAddressAllocator<uint32_t>>::value)
+		{
+			alctr = AlctrType(nullptr, randAllocParams.offset, randAllocParams.alignOffset, randAllocParams.maxAlign, randAllocParams.addressSpaceSize);
+		}
+		else
+		{
+			const auto reservedSize = AlctrType::reserved_size(randAllocParams.maxAlign, randAllocParams.addressSpaceSize, randAllocParams.blockSz);
+			reservedSpace = _NBL_ALIGNED_MALLOC(reservedSize, _NBL_SIMD_ALIGNMENT);
+			alctr = AlctrType(reservedSpace, randAllocParams.offset, randAllocParams.alignOffset, randAllocParams.maxAlign, randAllocParams.addressSpaceSize, randAllocParams.blockSz);
+		}
```
Comment (devshgraphicsprogramming) [inline examples_tests/10.AllocatorTest/main.cpp:130]: this should move to `executeAllocatorTest` 

because right now you're only testing `reset` at most once per allocator creation (you dont test reuse of allocator after reset)

File: examples_tests/43.AllocatorTest/main.cpp (hunk 8 above)
Comment (devshgraphicsprogramming) [reply examples_tests/43.AllocatorTest/main.cpp]: > random number of times you try different parameters (buffer size, etc.) to construct a new allocator

this would be the `executeAllocatorTest` being invoked multiple times! Not done!

Comment (devshgraphicsprogramming) [reply examples_tests/43.AllocatorTest/main.cpp]: > random number of iteration loops (times reset of an allocator is performed)

Not done because in the function that does the allocs + reset you create a new address allocator, but not far off

File: examples_tests/10.AllocatorTest/main.cpp [hunk 33]
```diff
@@ -0,0 +1,420 @@
...
+				alctr.reset();
+				results.clear();
+			}
+			else
+			{
+				// free everything with a series of multi_free
+				while (results.size() != 0u)
+					randFreeAllocatedAddresses(alctr);
+			}
+		}
```
Comment (devshgraphicsprogramming) [inline examples_tests/10.AllocatorTest/main.cpp:168]: there's no reset or free when you use linear address allocator ....

File: examples_tests/10.AllocatorTest/main.cpp [hunk 34]
```diff
@@ -0,0 +1,420 @@
...
+		 
+		if constexpr (!std::is_same<AlctrType, core::LinearAddressAllocator<uint32_t>>::value)
+			_NBL_ALIGNED_FREE(reservedSpace);
```
Comment (devshgraphicsprogramming) [inline examples_tests/10.AllocatorTest/main.cpp:171]: move that out to `executeAllocatorTest`

File: examples_tests/43.AllocatorTest/main.cpp (hunk 8 above)
Comment (devshgraphicsprogramming) [reply examples_tests/43.AllocatorTest/main.cpp]: > random number of multi-alloc calls with random number of allocations in the multi-alloc

Looks good and proper.

> random number of multi-free calls (not every multi-allocation is paired 1:1 with multi-free)

Looks good and proper

File: examples_tests/10.AllocatorTest/main.cpp (hunk 33 above)
Comment (Przemog1) [reply examples_tests/10.AllocatorTest/main.cpp:168]: it is always reseted here: https://github.com/Devsh-Graphics-Programming/Nabla/pull/1/files/e575e6a4745f77f2be5657dc667808d1a775d997#diff-fd4cbad7bd04514c211c9432df7dfbbf19bdb575c02570581349ac24ad31d2d8R150

Comment (devshgraphicsprogramming) [reply examples_tests/10.AllocatorTest/main.cpp:168]: `randFreeAllocatedAddresses` should not reset an allocator, also from what I see.. it only calls `multi_free` which for a linear allocator is a NOOP

Comment (Przemog1) [reply examples_tests/10.AllocatorTest/main.cpp:168]: there is template specialization for a linear allocator:
https://github.com/Devsh-Graphics-Programming/Nabla/pull/1/files/e575e6a4745f77f2be5657dc667808d1a775d997#diff-fd4cbad7bd04514c211c9432df7dfbbf19bdb575c02570581349ac24ad31d2d8R285

Comment (devshgraphicsprogramming) [reply examples_tests/10.AllocatorTest/main.cpp:168]: `randFreeAllocatedAddresses` should only allocate a subset of addresses... so for linear allocator that would be a "random" `reset` (with really low probability)

the true reset should happen at the end... here
